NSUUID with UUID::deletedValue is converted to a random WTF::UUID

szewai opened the pull request:
#### 424980e4bed28f494a3635d9c17e38ef8b0bd9cc
<pre>
NSUUID with UUID::deletedValue is converted to a random WTF::UUID
<a href="https://bugs.webkit.org/show_bug.cgi?id=253805">https://bugs.webkit.org/show_bug.cgi?id=253805</a>
rdar://106626402

Reviewed by Geoffrey Garen.

Add a new static function UUID::fromNSUUID to replace UUID(NSUUID *), which would return std::nullopt if NSUUID contains
UUID::deleteValue.

* Source/WTF/wtf/UUID.cpp:
(WTF::UUID::parse):
* Source/WTF/wtf/UUID.h:
* Source/WTF/wtf/cocoa/UUIDCocoa.mm:
(WTF::UUID::fromNSUUID):
(WTF::UUID::UUID): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(+[WKWebsiteDataStore dataStoreForIdentifier:]):
(+[WKWebsiteDataStore _removeDataStoreWithIdentifier:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerConfiguration.mm:
(+[_WKWebExtensionControllerConfiguration configurationWithIdentifier:]):
(-[_WKWebExtensionControllerConfiguration initWithCoder:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.mm:
(-[_WKWebsiteDataStoreConfiguration initWithIdentifier:]):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WTF/cocoa/UUIDCocoa.mm: Copied from Source/WTF/wtf/cocoa/UUIDCocoa.mm.
(TEST):

Canonical link: <a href="https://commits.webkit.org/261677@main">https://commits.webkit.org/261677@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1f60e140699c79a2f80918f4bd79cfdd6c69ee5c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112494 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21642 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1146 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4272 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121068 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/116560 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22980 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/12800 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/5411 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118261 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/17068 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100265 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105581 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/99023 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/815 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/46094 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/100815 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13987 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/848 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/95170 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/12093 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14679 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/10227 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/102271 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20018 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52853 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/31946 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8135 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16503 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/110310 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/27248 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WTF/wtf/cocoa/UUIDCocoa.mm:
```diff
@@ -35,11 +35,17 @@
-UUID::UUID(NSUUID * nsUUID)
-    : UUID(*parse(String([nsUUID UUIDString])))
+UUID::UUID(NSUUID *nsUUID)
+    : m_data(deletedValue)
```
I see a few potential problems here:

1. Default initialization to the deleted value is probably incorrect. Usually we default initialize to the empty value. One thing that will be incorrect is operator bool, which checks for the empty value. And logically, a default UUID is not a deleted bucket in a hash table.

2. Why is it OK for some operations to fail if the client happens to use the 0x0 or 0x1 UUID? (A client generating random UUIDs will never hit this case; but a client making them by hand might.) Many WebCore data types use special empty and deleted values, and we make that work by controlling all the values ourselves. But UUIDs can come from uncontrolled places, like public APIs.

3. What guarantees that all clients of the UUID class will know to check for validity before using?

I think it might be better for the UUID class not to offer a hash table emptyValue or deletedValue. Instead, clients that want to put a UUID in a hash table should use a wrapper that can represent empty and deleted. (Maybe we already support this using optional or pair?)

geoffreygaren replied:
Sorry these comments are late. I didn't push the publish button.

szewai replied:
> Default initialization to the deleted value is probably incorrect. Usually we default initialize to the empty value. One thing that will be incorrect is operator bool, which checks for the empty value. And logically, a default UUID is not a deleted bucket in a hash table.

Sure, I will change it to empty value.

> Why is it OK for some operations to fail if the client happens to use the 0x0 or 0x1 UUID? (A client generating random UUIDs will never hit this case; but a client making them by hand might.) Many WebCore data types use special empty and deleted values, and we make that work by controlling all the values ourselves. But UUIDs can come from uncontrolled places, like public APIs.

0 is nil UUID (https://www.ietf.org/rfc/rfc4122.txt), so it's a special value by spec and might be okay to fail the operations. 
1 is should be a valid value; I think it might be unexpected to fail operations on it.

> What guarantees that all clients of the UUID class will know to check for validity before using

I think identifier classes (e.g. `ObjectIdentifier`) generally need to be validated before using?

> I think it might be better for the UUID class not to offer a hash table emptyValue or deletedValue. Instead, clients that want to put a UUID in a hash table should use a wrapper that can represent empty and deleted. (Maybe we already support this using optional or pair?)

Maybe we could introduce a new class RandomUUID for existing use cases, which has emptyValue and deletedValue (i.e. cannot 0 or 1) and can be used for hash table. And make UUID accept 0 and 1 (mapping to value in NSUUID) which cannot be used for hash table directly?
(I am not sure what you mean by supporting using optional or pair...)

This patch just fixes the bug that NSUUID of 0x1 is converted to a WTF::UUID with random value (i.e. NSUUID -> WTF::UUID -> different NSUUID). I think "whether 0x0 or 0x1 should be treated as valid value" is a different issue that needs more discussion.

Constellation replied:
I think we should validate and reject deletedValue NSUUID. Instead of having a constructor, UUID::fromNSUUID should exist and it should return std::optional based on the value. And when 0x1 NSUUID comes, we should just reject it.

The deletedValue and emptyValue are not selected randomly in UUID value space. Since canonical UUID needs version header (see the spec of UUID), these two values are never valid UUID. UUID generation must not generate these values (otherwise, it is not a UUID). So from the input, when accepting it as WTF::UUID, we should reject them. If they are generated in a right manner, they are never 0x1. And NSUUID already rejects 0x0 UUID (it is also the invalid value). So it is natural that we reject the other invalid UUIDs.

I prefer current deletedValue / emptyValue largely because of memory reason. Having another boolean for HashMap / HashSet is too costly since UUID is 16-byte aligned: This means these boolean gets 16-byte size.

Constellation replied:
BTW, we sometimes would like to use all the value range of the keys in HashMap. For that purpose, we already have GenericHashKey<T> wrapper. See some use cases in JSC. But for UUID, I think rejecting invalid UUID when adopting WTF::UUID looks a right path to me.

geoffreygaren replied:
Thanks! FWIW, GenericHashKey is what I was trying to find when I suggested

> I think it might be better for the UUID class not to offer a hash table emptyValue or deletedValue. Instead, clients that want to put a UUID in a hash table should use a wrapper that can represent empty and deleted. (Maybe we already support this using optional or pair?)

Given that @Constellation and @szewai have pointed out that both 0x0 and 0x1 are defined to be invalid UUIDs by spec, I like the suggestion to
* change UUID(NSUUID*) to UUID::fromNSUUID, and have it return optional (so we can check for invalid input)
* always fail UUID parsing if 0x0 or 0x1 is the input (prevents hash table corruption, clearer)
* keep using emptyValue 0x0 and deletedValue 0x1 (saves memory, simpler)

If there is a use case for 0x0 and 0x1 that we know we use, I would also be OK with a GenericHashKey solution or with having two classes, UUID and UUIDWithRFCIncompatibleBytes. But that seems more complicated.

szewai replied:
According to spec (https://www.rfc-editor.org/rfc/rfc4122#section-4.1.2) the last byte is Node, not version, so I am not sure why 0x1 is picked.
I agree a normal UUID generator will not generate 0x0 or 0x1 though (based on the rules specified in the spec), so we might just reject it. The proposal sounds good!

Constellation replied:
> According to spec ([rfc-editor.org/rfc/rfc4122#section-4.1.2](https://www.rfc-editor.org/rfc/rfc4122#section-4.1.2)) the last byte is Node, not version, so I am not sure why 0x1 is picked.

I mean, UUID version number is 1-5. So if the whole UUID value is 0x1, it means version number is 0, thus, not a valid UUID.
https://www.rfc-editor.org/rfc/rfc4122#section-4.1.3

szewai replied:
I see. I think that means every value with version 0 should be treated as invalid (e.g. 0x2, 0x3, ...), but existing implementation actually allows those values, so the criteria might be confusing to users (who crafts NSUUID with special values): "why 1 is not valid in WebKit but 2 is"

geoffreygaren replied:
If it's clearer to be even more restrictive, I think that's an OK thing to do. Better to be restrictive up front. Once we start accepting UUIDs and putting them on disk, we probably can't go back.

youennf commented on Source/WebCore/Modules/notifications/NotificationDataCocoa.mm:
```diff
@@ -57,12 +57,12 @@
     String uuidString = dictionary[WebNotificationUUIDStringKey];
     auto uuid = UUID::parseVersion4(uuidString);
-    if (!uuid)
+    if (!uuid || !uuid->isValid())
```
I do not think this is needed, 0 and 1 should not be allowed values for UUID Version4, ditto below.

youennf commented on Source/WebKit/UIProcess/Cocoa/SystemPreviewControllerCocoa.mm:
```diff
@@ -361,7 +361,7 @@ - (UIImage *)previewController:(QLPreviewController *)controller transitionImage
     ASSERT(uuid);
-    if (!uuid)
+    if (!uuid || !uuid->isValid())
         return;
```
If we change UUID::parse(), we might want to look at WebsiteDataStore::fetchAllDataStoreIdentifiers.

szewai replied:
Updated patch does not change this behavior.

youennf commented on Source/WTF/wtf/cocoa/UUIDCocoa.mm:
```diff
@@ -35,11 +35,17 @@
+    // and we store UUID as UInt128).
+    auto uuid = parse(String([nsUUID UUIDString]));
+    ASSERT(uuid);
+    m_data = uuid->data();
```
I am not sure we want to be able to construct a deleted or empty value without noticing it. Currently, we can create such UUID values explicitly only.
Would it be better to have a static std::optional<UUID> fromNSUUID method, similarly to the parse methods?

szewai replied:
Technically NSUUID with WebCore::UUID::deletedValue is a valid uuid, so we probably should make UUID accept this value (not treating it as invalid).

youennf replied:
As it is, this change could easily allow WebKit map corruptions.
If we want to support 1 as a valid UUID, we should remove the deleted UUID value and update our WebKit code accordingly as a first step.

youennf commented on Tools/TestWebKitAPI/Tests/WTF/UUID.cpp:
```diff
@@ -95,3 +95,11 @@ TEST(WTF, TestUUIDVersion4MakeString)
+    String deletedValueString = "00000000-0000-0000-0000-000000000001"_s;
+    auto uuid = UUID::parse(deletedValueString);
+    EXPECT_TRUE(!!uuid);
+    EXPECT_FALSE(uuid->isValid());
```
By default, parse should return a valid value or std::null opt.
It might be fine to make parse() take a AllowInvalidValue boolean enum parameter (with a default value equal to No to not change the default behavior).
Only UID::UUID(NSUUID *nsUUID) would then call parse() with AllowDeletedValue::Yes.

szewai replied:
This proposal sounds good to me; I will update the patch

geoffreygaren commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerConfiguration.mm:
```diff
@@ -59,7 +59,9 @@ + (instancetype)configurationWithIdentifier:(NSUUID *)identifier
 
-    return WebKit::WebExtensionControllerConfiguration::create(identifier)->wrapper();
+    auto uuid = UUID::fromNSUUID(identifier);
+    ASSERT(uuid);
```
Probably this should be RELEASE_ASSERT, since the consequence may effect data on disk.

geoffreygaren commented on Source/WTF/wtf/cocoa/UUIDCocoa.mm:
```diff
@@ -35,11 +35,14 @@
+
+    // Use string instead of bytes to avoid consideration of endianess (NSUUID stores UUID as array of bytes,
+    // and we store UUID as UInt128).
+    return parse(String([nsUUID UUIDString]));
```
I think you also need to add an explicit check inside parse so that parsing fails for emptyValue. I guess we need an API test for that too.

geoffreygaren reviewed (approved):
r=me

szewai committed:
NSUUID with UUID::deletedValue is converted to a random WTF::UUID
https://bugs.webkit.org/show_bug.cgi?id=253805
rdar://106626402

Reviewed by Geoffrey Garen.

Add a new static function UUID::fromNSUUID to replace UUID(NSUUID *), which would return std::nullopt if NSUUID contains
UUID::deleteValue.

* Source/WTF/wtf/UUID.cpp:
(WTF::UUID::parse):
* Source/WTF/wtf/UUID.h:
* Source/WTF/wtf/cocoa/UUIDCocoa.mm:
(WTF::UUID::fromNSUUID):
(WTF::UUID::UUID): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(+[WKWebsiteDataStore dataStoreForIdentifier:]):
(+[WKWebsiteDataStore _removeDataStoreWithIdentifier:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerConfiguration.mm:
(+[_WKWebExtensionControllerConfiguration configurationWithIdentifier:]):
(-[_WKWebExtensionControllerConfiguration initWithCoder:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.mm:
(-[_WKWebsiteDataStoreConfiguration initWithIdentifier:]):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WTF/cocoa/UUIDCocoa.mm: Copied from Source/WTF/wtf/cocoa/UUIDCocoa.mm.
(TEST):

Canonical link: https://commits.webkit.org/261677@main
